### PR TITLE
Fixed #25947, #33330 -- Support printing query using non-default database.

### PR DESCRIPTION
**Pull request by tjni**
This allows writing "print(query_set.query)" when the query can only be
printed by a different database backend.

I also updated a few other places in Query that were using the default DB to reference the DB plumbed into the Query from the QuerySet.

~~A few quotes were also changed from single quote to double quote in the test file, due to running the whole file through the Black code formatter.~~

### CODE REVIEW

File: tests/queries/test_query.py [hunk 1]
```diff
@@ -140,21 +158,33 @@ def test_filter_conditional(self):
...
+        # the default DB to test that the method call works. We can fall back to
+        # checking that the correct connection is looked up and assume that it's
+        # used.
+        connections.__getitem__.assert_called_once_with(OTHER_DATABASE_ALIAS)
+
 
 class JoinPromoterTest(SimpleTestCase):
     def test_repr(self):
         self.assertEqual(
-            repr(JoinPromoter('AND', 3, True)),
```
Comment (felixxm) [inline tests/queries/test_query.py:158]: Please revert all unrelated changes from single to double quotes.

Comment (tjni) [reply tests/queries/test_query.py:158]: I can do this. For future reference, is this to keep the change specific, or is there a preference for single quotes?

Comment (felixxm) [reply tests/queries/test_query.py]: Currently both.

Comment (tjni) [reply tests/queries/test_query.py]: Cool, got it. Let me update the code I've written to use single quotes too then.

File: tests/multiple_database/tests.py [hunk 2]
```diff
@@ -110,7 +110,7 @@ def test_refresh_router_instance_hint(self):
         book = Book.objects.create(title='Dive Into Python', published=datetime.date(1957, 10, 12))
         with self.settings(DATABASE_ROUTERS=[router]):
             book.refresh_from_db()
-        router.db_for_read.assert_called_once_with(Book, instance=book)
+        router.db_for_read.assert_called_with(Book, instance=book)
```
Comment (felixxm) [inline tests/multiple_database/tests.py:113]: This is worrying :thinking:

**Review by felixxm — COMMENTED** [review]
@tjni Thanks for this patch :+1: However, I think it's not an acceptable solution. We cannot just remove a comment (which applies to all your changes from `DEFAULT_DB_ALIAS` to `self.using`) with a strong warning that _"... does not have (**nor should it have**) knowledge of which connection is going to be used."_. This behavior basically exists since we added multiple database support in ff60c5f9de3e8690d1e86f3e9e3f7248a15397c8, see also [discussion about API](https://groups.google.com/g/django-developers/c/CfA1P-BoK3M).

Moreover `Query.using` doesn't support `QuerySet.using()` so it's extremely confusing because it gives an impression that `using` is honored, but it's not, it will use `DEFAULT_DB_ALIAS` in most of cases, e.g.
```python
>>> qs = Author.objects.using('other')
>>> qs.db
other
>>> qs.query.using
default
```
Previous implementation was clear, we don't have (nor should have) a real connection so we need to use `DEFAULT_DB_ALIAS` now this logic is hidden. Taking into account that `QuerySet.query` is an internal API maybe it's enough to add the `using` argument to the `sql_with_params()` method :thinking: It would be ideal to have access to `QuerySet.db` in `Query.__str__()`, but ...

**Comment by tjni** [comment]
I really appreciate the time you've taken to explain your thinking, dig up the relevant discussion from years ago, and to provide an example of a confusing scenario (that I would classify as a bug if it were allowed).

What this makes clear is more discussion is needed at the conceptual level. I'm still new to this project, so can you suggest which is the best way to do this. Is it via comments in this PR, back on Trac, or some other method?

**Comment by felixxm** [comment]
> What this makes clear is more discussion is needed at the conceptual level. I'm still new to this project, so can you suggest which is the best way to do this? Is it via comments in this PR, back on Trac, or some other method?

It depends what do you want to achieve:
- if you want to discuss an alternative solution we can do this here, on the [django-developers mailing list](https://groups.google.com/g/django-developers), or on the [forum](https://forum.djangoproject.com/),
- if you disagree with my opinion that this solution is unacceptable, and you would like to move it forward then a discussion on the [django-developers mailing list](https://groups.google.com/g/django-developers), or on the [forum](https://forum.djangoproject.com/) would be appropriate (see [_"How Django is developed"_](https://github.com/django/deps/blob/main/accepted/0010-new-governance.rst#how-django-is-developed)).

Generally raising an idea on the [mailing list](https://groups.google.com/g/django-developers) or the [forum](https://forum.djangoproject.com/) is the best way to reach a wider audience and see what other think.

**Comment by tjni** [comment]
I agree with your opinion, so perhaps it would be practical to get your thoughts first, and you can let me know if seeking other perspectives feels appropriate.

I read the first couple of comments from the API discussion and skimmed through the rest, and my interpretation is that it is brainstorming the best way for users to influence database routing (e.g. through a manager, through a the query set, through a model's save method, etc.) It all reads very sensibly to me.

Based on this information and the comments in the code, I feel uncertainty about what a query _is_, so I'd like your thoughts on it. Is a query functional without a query set? Is a query set functional without a connection?

From my own perspective as an end user, when I print a query, I do not care about the abstract representation but the concrete one, so I wouldn't mind an opaque string when a connection is missing. When I print it through a query set that has been endowed with a connection, such as through a `using` call, I expect that the connection is inferred.

I am currently gaining access to the query that I need from `connection.queries`, but I do think it is less intuitive and friendly than if I can print the query directly.
